subsys/suit/stream/stream_sinks/src/suit_sdfw_recovery_sink.c

```diff
@@ -122,126 +77,154 @@ static suit_plat_err_t schedule_sdfw_recovery_update(const uint8_t *buf, size_t
+		break;
 	}
+	default: {
+		/* SecROM indicates error during update */
+		LOG_ERR("Update failure: %08x", update_status);
+		/* By design SDFW Recovery update error should not result in failing whole
+		 * installation.
+		 * Because of that, set specific error code instead of SUIT_PLAT_ERR_CRASH.
+		 */
+		err = SUIT_PLAT_ERR_SDRFW_FAILURE;
+		break;
+	}
+	}
+
+	return err;
 }
 
-static suit_plat_err_t check_urot_none(const uint8_t *buf, size_t size)
+static suit_plat_err_t sdfw_update_ongoing(const uint8_t *buf, size_t size)
 {
-	/* Detect update candidate. */
-	/* It is enough to check Public Key Size field which occupies first 4B of Signed Manifest.
-	 */
-	if (*((uint32_t *)buf) == EMPTY_STORAGE_VALUE) {
-		LOG_INF("Update candidate not found");
-		return SUIT_PLAT_ERR_NOT_FOUND;
+	suit_plat_err_t err = SUIT_PLAT_SUCCESS;
+
+	enum sdfw_update_status update_status = sdfw_update_initial_status_get();
+
+	/* SDFW update was ongoing. */
+	switch (update_status) {
+	case SDFW_UPDATE_STATUS_NONE: {
+		/* No pending operation even though operation indicates SDFW update.
+		 * Proceed with SDFW Recovery update.
+		 */
+		err = schedule_update_and_reboot(buf, size);
+		break;
+	}
+	case SDFW_UPDATE_STATUS_UROT_ACTIVATED: {
+		/* SDFW was installed successfully.
+		 * Proceed with SDFW Recovery update.
+		 */
+		err = schedule_update_and_reboot(buf, size);
+		break;
+	}
+	default: {
+		/* SecROM indicates error during SDFW update.
+		 * By design SDFW update error should abort installation.
+		 */
+		LOG_ERR("SDFW update failure: %08x", update_status);
+		err = SUIT_PLAT_ERR_CRASH;
+		break;
+	}
 	}
 
-	LOG_INF("Update candidate found");
+	return err;
+}
 
-	suit_plat_err_t err = check_update_candidate(buf, size);
+static suit_plat_err_t update_needed_action(const uint8_t *buf, size_t size)
+{
+	suit_plat_err_t err = SUIT_PLAT_SUCCESS;
 
-	if (err == SDFW_RECOVERY_SINK_ERR_AGAIN) {
-		/* Update scheduled, continue after reboot */
-		reboot_to_continue();
-		if (IS_ENABLED(CONFIG_SUIT_UPDATE_REBOOT_ENABLED)) {
-			/* If this code is reached, it means that reboot did not work. */
-			/* In such case report an error and convert the error code. */
-			LOG_ERR("Expected reboot did not happen");
-			err = SUIT_PLAT_ERR_UNREACHABLE_PATH;
-		}
+	enum sdfw_update_operation initial_operation = sdfw_update_initial_operation_get();
+
+	switch (initial_operation) {
+	case SDFW_UPDATE_OPERATION_NOP: {
+		/* No previously running update or after the other slot update. */
+		/* Schedule an update of this slot. */
+		err = schedule_update_and_reboot(buf, size);
+		break;
+	}
+	case SDFW_UPDATE_OPERATION_UROT_ACTIVATE: {
+		/* SDFW update was ongoing */
+		err = sdfw_update_ongoing(buf, size);
+		break;
+	}
+	case SDFW_UPDATE_OPERATION_RECOVERY_ACTIVATE: {
+		/* SDFW Recovery update was ongoing */
+		err = recovery_update_ongoing(buf, size);
+		break;
+	}
+	default: {
+		LOG_ERR("Unhandled operation: %08x", initial_operation);
+		err = SUIT_PLAT_ERR_CRASH;
```

Comment:
Based on diagrams - this should be a soft fail.

Reply:
Good catch - changed to match the documentation diagram.